finality-aleph/src/data_io/status_provider.rs

```diff
@@ -0,0 +1,397 @@
+use crate::data_io::proposal::{AlephProposal, AvailableKind, ProposalStatus};
+use crate::BlockHashNum;
+use log::{error, warn};
+use lru::LruCache;
+use sc_client_api::HeaderBackend;
+use sp_runtime::traits::{Block as BlockT, Header as HeaderT, NumberFor};
+use sp_runtime::{generic::BlockId, SaturatedConversion};
+use std::{marker::PhantomData, sync::Arc};
+
+const BLOCKS_CACHE_SIZE: usize = 2000;
+const PROPOSALS_CACHE_SIZE: usize = 2000;
+
+pub trait ProposalStatusProvider<B: BlockT> {
+    fn on_block_imported(&mut self, header: B::Header);
+
+    fn on_block_finalized(&mut self, block: BlockHashNum<B>);
+
+    fn is_block_imported(&mut self, block: &BlockHashNum<B>) -> bool;
+
+    fn get_finalized_at(&mut self, number: NumberFor<B>) -> Result<BlockHashNum<B>, ()>;
+
+    fn get_parent_hash(&mut self, block: &BlockHashNum<B>) -> Result<B::Hash, ()>;
+
+    // Default implementation is provided but can be overriden to enable caching.
+    fn get_status(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        old_status: Option<&ProposalStatus>,
+    ) -> ProposalStatus {
+        self.get_status_inner(proposal, old_status)
+    }
+
+    fn is_hopeless_fork(&mut self, proposal: &AlephProposal<B>) -> bool {
+        let bottom_num = proposal.number_bottom_block();
+        for i in 0..proposal.branch.len() {
+            if let Ok(finalized_block) =
+                self.get_finalized_at(bottom_num + <NumberFor<B>>::saturated_from(i))
+            {
+                if finalized_block.hash != proposal.branch[i] {
+                    return true;
+                } else {
+                    // We don't know the finalized block at this height
+                    break;
+                }
+            }
+        }
+        false
+    }
+
+    fn is_ancestor_finalized(&mut self, proposal: &AlephProposal<B>) -> bool {
+        let bottom = proposal.bottom_block();
+        let parent_hash = if let Ok(hash) = self.get_parent_hash(&bottom) {
+            hash
+        } else {
+            return false;
+        };
+        let finalized = if let Ok(hash) = self.get_finalized_at(proposal.number_below_branch()) {
+            hash
+        } else {
+            return false;
+        };
+        parent_hash == finalized.hash
+    }
+
+    fn is_branch_correct(&mut self, proposal: &AlephProposal<B>) -> bool {
+        let bottom_num = proposal.number_bottom_block();
+        for i in 1..proposal.branch.len() {
+            let curr_block = (
+                proposal.branch[i],
+                bottom_num + <NumberFor<B>>::saturated_from(i),
+            )
+                .into();
+            match self.get_parent_hash(&curr_block) {
```

Comment:
this method indeed checks that the hashes are in parent-child relation, but unless `get_parent_hash((num, hash))` checks that `num` and `hash` are referring to the same block, we still do not have guarantee that the whole proposal is correct -- we are missing `number` check

Reply:
You are right. I changed the name of the method to `is_branch_ancestry_correct` and added a few comments. Since before this check we call `is_block_imported(&top_block)` (which literally checks that (hash, num) has been imported, so if there is `hash` but with different num, then it outputs `false`), the overall correctness of `get_status_inner` is OK.